classes/controller/admin/form.ctrl.php

```diff
@@ -490,4 +490,129 @@ protected function getInputFieldData($name = 'fieldData')
 
         return $fieldData;
     }
+
+    public function action_duplicate($id = null)
+    {
+        try {
+            /**
+             * @var $form Model_Form
+             */
+            $form = $this->crud_item($id);
+            if (empty($form->id)) {
+                throw new \Exception(__('The form was not found.'));
+            }
+            $context = $form->form_context;
+            static::duplicateForm($form);
+            // Send response
+            \Response::json(array(
+                'dispatchEvent' => array(
+                    'name' => 'Nos\Form\Model_Form',
+                    'action' => 'insert',
+                    'context' => $context,
+                ),
+                'notify' => __('Here you are! The form has just been duplicated.'),
+            ));
+        } catch (\Exception $e) {
+            $this->send_error($e);
+        }
+    }
+
+    /**
+     * @param Model_Form $form : the form to duplicate
+     * @throws \Exception
+     */
+    protected static function duplicateForm(Model_Form $form)
+    {
+        $clone = clone $form;
+        $try = 1;
+        do {
+            try {
+                $title_append = strtr(__(' (copy {{count}})'), array(
+                    '{{count}}' => $try,
+                ));
+                $clone->form_virtual_name = null;
+                $clone->form_name = $form->title_item().$title_append;
+                if ($clone->save()) {
```

Comment:
is this going to overwrite the previous form as you didn't reset the primary key ?